source/dub/project.d

```diff
@@ -353,14 +353,21 @@ class Project {
 						continue;
 					}
 				} else if (m_selections.hasSelectedVersion(basename)) {
-					vspec = m_selections.getSelectedVersion(basename);
-					if (vspec.path.empty) p = m_packageManager.getBestPackage(dep.name, vspec);
+					auto selection = m_selections.getSelectedVersion(basename);
+					if (selection.path.empty) p = m_packageManager.getBestPackage(dep.name, selection);
 					else {
-						auto path = vspec.path;
+						auto path = selection.path;
 						if (!path.absolute) path = m_rootPackage.path ~ path;
 						p = m_packageManager.getOrLoadPackage(path, NativePath.init, true);
 						if (subname.length) p = m_packageManager.getSubPackage(p, subname, true);
+						// submodule folders checked out to a branch match any version (convenience)
+						if (p.version_.isBranch) selection = Dependency(m_rootPackage.path ~ selection.path);
+						else selection = Dependency(p.version_, m_rootPackage.path ~ selection.path);
 					}
+					vspec = vspec.merge(selection);
+					enforce(vspec != Dependency.invalid, format(
+						"Selected package %s doesn't match %s(%s) of %s",
+							selection, dep.name, dep.spec, pack.name));
```

Comment:
If I understand this correctly, this will probably break the way I use Dub + submodules where I bind-mount the submodule to its "real" location (i.e. `bindfs -n ~/work/ae lib/ae`) - since the main directory is on master, it won't match the `dub.selections.json` version. That's probably fine though. If there isn't a way to override it (local `dub.selections.json` override?) then maybe it would be nice if there was a way.

Reply:
Commit message should conform to git conventions (first line under 60-80 characters).

Reply:
Should work. Branches are excluded from the override, for pretty much that reason that it would make incremental development impossible, which is exactly the thing submodules/bind mounts are for.